### ISSUE TITLE
Feature Request: unwind function.

### ISSUE THREAD

**Issue by DevBaweja** [ramda/ramda#3078]
## 1) Design

```
Deconstructs an array field from the input objects to output a object for each element.
Each output object is the input object with the value of the array field replaced by the element.
@since version
@category Array
@param {array} The array of objects which you want to unwind.
@param {key} Key of object containing the array to be unwind.
```

## 2) Output

-   ### Basic

```
unwind([
    { name: 'Francisco', hobbies: ['Hacking', 'Ice skating'] },
    { name: 'Adelina', hobbies: ['Juggling', 'Graffiti', 'Painting'] },
], 'hobbies');
=> [
    { name: 'Francisco', hobbies: 'Hacking'},
    { name: 'Francisco', hobbies: 'Ice skating'},

    { name: 'Adelina', hobbies: 'Juggling'},
    { name: 'Adelina', hobbies: 'Graffiti'},
    { name: 'Adelina', hobbies: 'Painting'},
]
```

Inform us if you also want us to implement these ideas.

**Comment by Bradcomp** [comment]
This seems like a pretty reasonable function to add. It has clear semantics and use cases. We should swap the argument order so the property to unwind comes before the data to be unwinded.

Are you interested in opening a PR?

**Comment by CrossEye** [comment]
I agree that this is interesting, but before creating a PR, I think we should also look at a possible expansion of the idea:

```js
unwind (['hobbies', 'colors'], [
    { name: 'Francisco', hobbies: ['Hacking', 'Ice skating'], colors: ['red', 'blue'] },
    { name: 'Adelina', hobbies: ['Juggling', 'Graffiti', 'Painting'], colors: ['green'] },
]);
// => [
//     { name: 'Francisco', hobbies: 'Hacking', colors: 'red'},
//     { name: 'Francisco', hobbies: 'Ice skating', colors: 'blue'},
//     { name: 'Francisco', hobbies: 'Hacking', colors: 'red'},
//     { name: 'Francisco', hobbies: 'Ice skating', colors: 'blue'},
//     { name: 'Adelina', hobbies: 'Juggling', colors: 'green'},
//     { name: 'Adelina', hobbies: 'Graffiti', colors: 'green'},
//     { name: 'Adelina', hobbies: 'Painting', colors: 'green'},
// ]
```

Or something like

```js
unwind ({hobby: 'hobbies', color: 'colors'}, [
    { name: 'Francisco', hobbies: ['Hacking', 'Ice skating'], colors: ['red', 'blue'] },
    { name: 'Adelina', hobbies: ['Juggling', 'Graffiti', 'Painting'], colors: ['green'] },
]);
// => [
//     { name: 'Francisco', hobby: 'Hacking', color: 'red'},
//     { name: 'Francisco', hobby 'Ice skating', color: 'blue'},
//     { name: 'Francisco', hobby: 'Hacking', color: 'red'},
//     { name: 'Francisco', hobby: 'Ice skating', color: 'blue'},
//     { name: 'Adelina', hobby: 'Juggling', color: 'green'},
//     { name: 'Adelina', hobby: 'Graffiti', color: 'green'},
//     { name: 'Adelina', hobby: 'Painting', color: 'green'},
// ]
```

For that latter one I'm not sure whether the configuration should be `{hobbies: 'hobby', colors: 'color'}` or `{hobby: 'hobbies', color: 'colors'}`.  Either would probably do.

Or if we liked the key renaming but not the multiple occurrences, we could just do `unwind ('hobbies', 'hobby', xs)`.

----------

We should also consider what happens if `hobbies` is not an array.  Is this just an error, or do we try to do something more sophisticated?  I'm afraid of

```js
unwind ('hobbies', [
    { name: 'Francisco', hobbies: 'Golf'},
]);
// => [
//     { name: 'Francisco', hobby: 'G'},
//     { name: 'Francisco', hobby: 'o'},
//     { name: 'Francisco', hobby: 'l'},
//     { name: 'Francisco', hobby: 'f'},
// ]
```

which might just fall out of an obvious implementation.  But we might just treat that as Garbage In/ Garbage Out.

**Comment by DevBaweja** [comment]
@Bradcomp

Yes, I and my team are really interested in contributing to your repository. We will be creating a PR soon.
And for currying, argument will be swapped so that data comes last.

**Comment by DevBaweja** [comment]
@CrossEye

We will be creating another prototype containing all the test case and
all possible use of the unwind function.

**Comment by DevBaweja** [comment]
## Possible Ways of implementing unwind

### 1) Taking arrayOfKeys as first argument

```
@param {arrayOfKeys} Array of keys which is to be unwind.
@param {arrayOfObjects} Array of objects on which unwind will occur.

unwind(['hobbies', 'colors'],[
    { hobbies: ['h1', 'h2'], colors:['c1', 'c2'] },
]);
=> [
    { hobbies: 'h1', colors: 'c1'},
    { hobbies: 'h1', colors: 'c2'},

    { hobbies: 'h2', colors: 'c1'},
    { hobbies: 'h2', colors: 'c2'},
]
```

Output will contain all the combination of h1,h2 and c1,c2.
However, keys name remain the same. And renaming from hobbies -> hobby, colors -> color
can be done afterwards.

### 2) Taking ObjectsWithKeyAndRename as first argument

```
@param {ObjectsWithKeyAndRename} Object where key is key in final output and value is used to unwind the object.
@param {arrayOfObjects} Array of objects on which unwind will occur.

unwind({ hobbies: 'hobby', colors: 'color' },[
    { hobbies: ['h1', 'h2'], colors:['c1', 'c2'] },
]);
=> [
    { hobby: 'h1', color: 'c1'},
    { hobby: 'h1', color: 'c2'},

    { hobby: 'h2', color: 'c1'},
    { hobby: 'h2', color: 'c2'},
]
```

-   [x] { hobby: 'hobbies', color: 'colors' }
-   [ ] { hobbies: 'hobby', colors: 'color' }

If we consider this, user will expect same structure { hobby: ' ', color: ' ' } in output as given in input argument.

-   [ ] { hobby: 'hobbies', color: 'colors' }
-   [x] { hobbies: 'hobby', colors: 'color' }

If we consider this, it signifies that hobbies will be replaced by hobby which we think is more consistent.

### 3) Only Renaming Keys and not multiple occurrence

```
@param {key} key is used to unwind the object.
@param {keyRename} key is used in final output.
@param {arrayOfObjects} Array of objects on which unwind will occur.

const xs = [
    { hobbies: ['h1', 'h2'], colors:['c1', 'c2'] }
];

const ys = unwind('hobbies', 'hobby', xs);
=> [
    { hobby: 'h1', color: ['c1', 'c2']},
    { hobby: 'h2', color: ['c1', 'c2']},
]

const zs = unwind('colors', 'color', ys);

=> [
    { hobby: 'h1', color: 'c1'},
    { hobby: 'h1', color: 'c2'},

    { hobby: 'h2', color: 'c1'},
    { hobby: 'h2', color: 'c2'},
]
```

This is really the way to go. It avoid confusion and have very clear syntax.

### String as value to be unwind

```
unwind (hobbies, 'hobby', [
    { hobbies: 'Golf', colors: ['c1', 'c2']},
]);
=> [
    { hobby: 'Golf', color: ['c1', 'c2']},
]
```

In this case it will simply return the original value of key and will not further unwind the string

Let me know which of these you think will be best to implement for unwind function.
And we will sharing some more test cases of that specific case.

**Comment by CrossEye** [comment]
How to rename
-------------

If we go with the renaming on the fly (and I'm really not sure; I need to think on that some more), I think my preference would be for an API with `{hobbies: 'hobby', colors: 'color'}` as input.  It's only a minor preference for this over `{hobby: 'hobbies, color: 'colors'}`, but it has some nice consistency with renaming variables during object destructuring.

Should we rename?
-----------------

The downside of the renaming inside the function is that it feels a little too complex.  The function does too much for Ramda's usual simple API.

The upside is clear: it's definitely a more useful function this way.  For English speakers -- and I imagine it's the same in many other languages -- the use of the plural "hobbies" feels quite odd tied to the singular "Juggling".  If we don't include it here, I would want to also add a `mapKeys` / `renameKeys` or similar function, and would expect to usually use the first in conjunction with the second:  `const transform = pipe (unwind (['hobbies']), mapKeys ({hobbies: 'hobby}))`.  If most users would feel that way, then it makes little sense *not* to include it when the alternative behavior is not hard to specify using it:  `unwind({hobbies: 'hobbies'})`.

But remember than generally in Ramda API decisions, simplicity wins.


Apply to single value
---------------------

And that brings up another point.  I think this would be better applied to a *single* value and not an array of values.

```hs
unwind :: Dictionary -> Record -> [Record]
```

instead of the current suggestion of 

```hs
unwind :: Dictionary -> [Record] -> [Record]
```

It's easy enough to combine these, with `R.chain` or `Array.prototype.flatMap`.  For instance,

```js
unwind ({hobbies: 'hobby'}) (obj)
// and
chain (unwind ({hobbies: 'hobby'})) ([obj1, obj2, obj3])
// or
[obj1, obj2, obj3] .flatMap (unwind ({hobbies: 'hobby'}))
```

Handling unexpected properties
------------------------------

As to the non-array properties, I would probably simply not do any processing on that node.  So, for instance:

```js
unwind ({hobbies: 'hobby', colors: 'color'}) ({ name: 'Francisco', hobbies: 'Golf', colors: ['red', 'blue'] })
//=> [
//   { name: 'Francisco', hobbies: 'Golf', color: 'red' }
//   { name: 'Francisco', hobbies: 'Golf', color: 'blue' }
// ] 
```

Here we did nothing with "hobbies", not even renaming, because it wasn't an array.  This has nothing to do with it being a String.  We would do the same for any non-array value (maybe one day any non-iterable value, but not yet.)

Other Feedback?
---------------

Remember that this is one person's feedback. 

I would love to hear from others on this.  Does anyone have objections to this function?  Do you have any input on the details of the API discussed here?  Other thoughts?  @ramda/core.

**Comment by DevBaweja** [comment]
## Simplicity is best

We think that it is bit unnecessary to rename the keys in unwind function.
which brings us to have another function `mapKeys`.(discussed later)

About single value we suggest that design of function must be refactored.

```js
unwind :: Dictionary -> Record -> [Record]
```

is more useful.

```js
// @param {arrayOfKeys} The array of keys which you want to unwind.
// @param {object} object to be unwind.

unwind(['hobbies', 'colors'], {
    name: 'Alex',
    hobbies: ['Golf', 'Hacking'],
    colors: ['red', 'green']
});
=> [
    { name: 'Alex',hobbies: 'Golf', colors: 'red'},
    { name: 'Alex',hobbies: 'Golf', colors: 'green'},
    { name: 'Alex',hobbies: 'Hacking', colors: 'red'},
    { name: 'Alex',hobbies: 'Hacking', colors: 'green'},
]
```

It will support `multiple arrays` and create combination of elements of those arrays.
Instead of `[Record]` we now pass only `Record` as argument

And it will return `[Record]` with each object containing one combination of total combinations.

And for previous implementation where we take `[Record]` instead of `Record` can be done by chain on this newly created unwind as suggested.

## mapKeys / renameKeys function

Now, we come to renaming of the keys in `[Record]`.
There are two ways -

-   We create function that takes only `Record`. Named as renameKeys

```js
// { k: 'v' }
// rename k to v in object passed
renameKeys({ hobbies: 'hobby', colors: 'color' }, { name: 'Alex', hobbies: 'Golf', colors: 'red' });
// => { name: 'Alex', hobby: 'Golf', color:'red' }
```

This is useful anyways.

And in our case, we map over array.

```js
// unwind returns us [Record]
// we map over [Record] and apply renameKeys to each and every Record
pipe(unwind(['hobbies', 'colors']), map(renameKeys({ hobbies: 'hobby', colors: 'color' })));
```

-   Or, We create function that takes `[Record]`. Named as mapKeys

```js
const xs = [
    { name: 'Alex', hobbies: 'Golf', colors: 'red' },
    { name: 'Alex', hobbies: 'Golf', colors: 'green' },
    { name: 'Alex', hobbies: 'Hacking', colors: 'red' },
    { name: 'Alex', hobbies: 'Hacking', colors: 'green' },
];
mapKeys({ hobbies: 'hobby', colors: 'color' }, xs);

// => [
//     { name: 'Alex', hobby: 'Golf', color: 'red' },
//     { name: 'Alex', hobby: 'Golf', color: 'green' },
//     { name: 'Alex', hobby: 'Hacking', color: 'red' },
//     { name: 'Alex', hobby: 'Hacking', color: 'green' },
// ]
```

-   We were also thinking of mapKeys(function, xs) which allows us to dynamically name key of object in function.

For ex -

```js
const xs = [
    { ids: 'id1', name: 'Bob' },
    { ids: 'id2', name: 'Alex' },
];
mapKeys(
    applySpec({
        ids: prop('ids'),
    }),
    xs
);
// => [{ id1: 'id1', name: 'Bob' }, { id2: 'id2', name:'Alex' }];
```

## About Error Handling

Apologize for using String. We were being extra specific about that example only.

```js
unwind({ hobbies: 'hobby', colors: 'color' })({ name: 'Francisco', hobbies: 'Golf', colors: ['red', 'blue'] });
//=> [
//   { name: 'Francisco', hobbies: 'Golf', color: 'red' }
//   { name: 'Francisco', hobbies: 'Golf', color: 'blue' }
// ]
```

Non-iteratable values will be ignored.

We will wait untill some discussions is done on these functions. Once confirmed, if any of these is accepted. We will be producing some more test cases and proceed to implementing them.

**Comment by Bradcomp** [comment]
> Should we rename?

I don't think renaming is necessary. It makes more sense to me to do that separately. It seems to me to be a completely separate concern. Who is to say that the aggregated field for the `hobby` column in the database isn't just named `hobby` even though it's an Array. If it is named `hobbies` then maybe rename it before calling `unwind`, or after calling it.

I also think that using an Object as the first argument confuses things and could really create more edge cases 
```
unwind({});
unwind({ hobbies: undefined });
unwind({
 hobbies: 'hobby',
 toString() { return 'Blah'; }
});
```

> Taking arrayOfKeys as first argument

I can see the value in this, but it does add complexity, and is subject to some of the same edge cases, but I think `Array String` is at least a more reasonable signature. 

Whenever we have an array of inputs we run into the case of what to do if most of the inputs are good but some are bad.

What about doing the single property unwind, then an `unwindAll` that takes an array of keys?

Otherwise, couldn't we just `chain` multiple invocations to get it all unwinded?
```
pipe(
unwind('hobbies'),
chain(unwind('colors'))
)
```
I acknowledge there is a bit of a performance hit to doing it that way, but I don't know how major that would be. I wouldn't argue if the signature took an array of keys though. If most others think that more useful I will fall in line.

> Error Handling

I think it's reasonable to throw if the field being pointed to isn't an Array. I also think GIGO is a fine approach to take, but the check would be easy to make and not really effect that much at this point.

> mapKeys / renameKeys

While we don't have these functions in Ramda, there are sample implementations for both in the [cookbook](https://github.com/ramda/ramda/wiki/Cookbook#map-keys-of-an-object-rename-keys-by-a-function) and there are also some extension libraries that include them (linked in the cookbook).

**Comment by CrossEye** [comment]
@DevBaweja:

I'd love to hear more about your company and what brought you to working on Ramda.  You've already presented several very good ideas, and I'm excited to see them move forward.

> Apologize for using String. We were being extra specific about that example only.

Please, no apologies are necessary here.  I absolutely love your technique of presenting these ideas.  At this phase, while we're playing around with many variants of an API, we're all going to make some unjustified assumptions or oversimplified examples.  I was the one who brought up the `'Golf' ->['G', 'o', 'l', 'f']` possibility in the first place, mostly because there are a number of places where Ramda accidentally treats strings as arrays.  It was only later that I thought about other possibilities that couldn't even be iterated.

>We will wait untill some discussions is done on these functions. Once confirmed, if any of these is accepted. We will be producing some more test cases and proceed to implementing them.

Sounds great!  While I like to work in a test-driven fashion as much as I reasonably can, please don't feel it necessary to present your test cases for approval before submitting an implementation.  The norm here is to make a PR that is intended to be complete, adding the new function and test cases together with a change to `index.js`.  On the other hand, it it would make you more comfortable to proceed that way, there are certainly no objections to hashing out the test cases first.

----------

@Bradcomp:

> I don't think renaming is necessary. It makes more sense to me to do that separately. 

I agree.  I've needed a function like this several times, and think it would be worth having, but I don't think it belongs in Ramda.  As everyone has noted, it's easy enough to build out of the parts we're proposing.

And perhaps even worse than the cases you mention might be

```js
unwind ({hobbies: 'foo', colors: 'foo'})
```

What in the world would be the logical result of that?  

> Whenever we have an array of inputs we run into the case of what to do if most of the inputs are good but some are bad.

I'm not sure we could have any particularly bad input here if we're expecting an array of strings (or strings/numbers/symbols).  If the target key is not a property of the source object, I would expect to result is no error at all; we simply return an array containing just the original object.

But still, I agree that simplicity should rule here as well, and we should just take a single key to expand.  Later we could add a function that takes an array of them.

> I think it's reasonable to throw if the field being pointed to isn't an Array.

To me that just becomes a no-op, or rather just the wrapping of the original input in an array.

```js
unwind ('hobbies') ({name: 'Fred', pastimes: ['Juggling', 'Sleeping']})
//=> [{name: 'Fred', pastimes: ['Juggling', 'Sleeping']}]

unwind ('hobbies') ({name: 'Wilma', hobbies: 'Golf'})
//=> [{name: 'Wilma', hobbies: 'Golf'}]

unwind ('hobbies') ({name: 'Barney', hobbies: NaN})
//=> [{name: 'Wilma', hobbies: NaN}]
```

I don't see a reason for an exception or even any implementation-specific garbage.  These samples simply feel like the right answer to me, and they should be easy to achieve.

>> mapKeys / renameKeys

> While we don't have these functions in Ramda, there are sample implementations for both in the [cookbook][cb] and there are also some extension libraries that include them (linked in the cookbook).

I'm inclined at this point to include this function.  I personally have had several times recently where I needed it, and I think it will pair well with `unwind` in several useful ways.  But I think we're going to need to go through the same API exercise for that one.


  [cb]: https://github.com/ramda/ramda/wiki/Cookbook

**Comment by Bradcomp** [comment]
I think at this point I would agree with everything you laid out there in terms of the API of this function.

**Comment by DevBaweja** [comment]
## Refactored API

```js
newUnwind :: String -> Record -> [Record]
```

(We are not suggesting renaming the function newUnwind/oldUnwind. It is just for understanding.)
Let us make this really simple.

```js
// @param {key} key by which object will be unwinded
// @param {object} object to be unwinded
newUnwind('hobbies', {
    name: 'alice',
    hobbies: ['Golf', 'Hacking'],
    colors: ['red', 'green'],
});
// => [
//      { name: 'alice', hobbies: 'Golf', colors: ['red', 'green'] },
//      { name: 'alice', hobbies: 'Hacking', colors: ['red', 'green'] }
// ]
```

Also as everyone agreed that renaming is a separate concern. So, we will be focusing on String/List as the first argument instead of a Dictionary.
It can serve all the possible cases we have discussed so far.
For instance:-

```js
oldUnwind :: List -> [Record] -> [Record]
```

1. About `List` to `String` in the first argument

The same effect of List can be easily obtained by chaining.

2. About `[Record]` to `Record` in second argument
   We can apply our newly created newUnwind in map to each and every
   Record.
   Previously,

```js
const xs = [
    { name: 'Francisco', hobbies: ['Hacking', 'Ice skating'] },
    { name: 'Adelina', hobbies: ['Juggling', 'Graffiti', 'Painting'] },
];
oldUnwind(['hobbies'], xs);
// => [
//     { name: 'Francisco', hobbies: 'Hacking'},
//     { name: 'Francisco', hobbies: 'Ice skating'},

//     { name: 'Adelina', hobbies: 'Juggling'},
//     { name: 'Adelina', hobbies: 'Graffiti'},
//     { name: 'Adelina', hobbies: 'Painting'},
// ]
```

Now,

```js
pipe(map(newUnwind(['hobbies']))(xs);
// => [
//     [
//         { name: 'Francisco', hobbies: 'Hacking'},
//         { name: 'Francisco', hobbies: 'Ice skating'},
//     ],
//     [
//         { name: 'Adelina', hobbies: 'Juggling'},
//         { name: 'Adelina', hobbies: 'Graffiti'},
//         { name: 'Adelina', hobbies: 'Painting'},
//     ]
// ]
```

Yes, there is slight differences in final output structure.
`[{},{},{},{},{},{}] is replaced with [ [{},{}], [{},{},{}]]`
As newUnwind combined the Record with name: 'Francisco' | name: 'Adelina' in its own list which is quite correct in some sense.
Although it can be solved by flattening the list.

## So, what's next?

```js
unwind :: String -> Record -> [Record]
```

is the simplest way to implement it with clear syntax and beginner's friendly nature.

> What about doing the single property unwind, then an unwindAll that takes an array of keys?

In, future we can implement all sort of functions related to unwind

Ex:-

```js
unwindAll :: List -> Record -> [Record]
unwindAll(['hobbies', 'colors'], xs);
```

Also, some sort of the inverse of these functions

```js
wind :: String -> [Record] -> Record
wind('hobby',[
    { name: 'alice', hobby: 'Hacking' },
    { name: 'alice', hobby: 'Juggling' },
])
// => { name: 'alice', hobby: ['Hacking', 'Juggling'] }
```

@CrossEye

> I agree. I've needed a function like this several times,
> and think it would be worth having, but I don't think it belongs in Ramda.
> As everyone has noted, it's easy enough to build out of the parts we're proposing.

> I'm inclined at this point to include this function.
> I think it will pair well with unwind in several useful ways.

We also agree it is quite trivial to build these functions.`(mapKeys/renameKeys)`
However, we also think that these functions have the potential to be included in ramda.
That can be our discussion for some other time.

> I'd love to hear more about your company and what brought you to working on Ramda

We are a group of students currently pursuing B.Tech in Computer Science. We want to contribute to an open-source project. And we think ramda is a great way to do so due to its simple architecture and efficiency. We are thankful that you approved our request. Looking forward to work and coordinate on the Ramda Project with you.

**Comment by CrossEye** [comment]
@DevBaweja:

I agree with most everything in your recent message but want to make one minor correction and one observation.

Correction: What I said does not belong in Ramda was what might be considered the largest and most involved version we discussed, one which combines the renaming with the unwinding and applies to one or more array properties.  That's too specialized and can easily be built out of the things we *are* talking about doing, `unwind` and `mapKeys/renameKeys`.

Observation.  In this example:

```js
pipe(map(newUnwind(['hobbies']))(xs);
```

the `pipe` call is unnecessary, And `map` can be replaced with `chain` to get a flattened list.

----------

Also, welcome!

> I'd love to hear more about your company and what brought you to working on Ramda

>> We are a group of students currently pursuing B.Tech in Computer Science. We want to contribute to an open-source project. 

We're very glad to have your help.  Are you classmates at the same university?

----------

Finally, it doesn't sound as though we're going to hear real objections to this from anyone.  So please, feel free to create a PR for this function.

**Comment by CrossEye** [comment]
@DevBaweja: Any updates on this?  I would love to see a PR,

**Comment by koc-299792458** [comment]
@CrossEye: Sorry for late reply but we are a group of students, including @DevBaweja working under a coach, so this gonna be a little bit slower as we are working according to coach as we are taking courses from him we will be opening PR soon.

**Comment by KayaLuken** [comment]
I think making `unwind` take an object to rename keys is overkill. Its a niche function to begin with. Even making it accept an array is borderline.

Or you could at least have counterpart `unwind` and `unwindAll` functions to handle those 2 cases.

**Comment by koc-299792458** [comment]
We are ready with final planing soon you gonna see PR in week or less

**Comment by DevBaweja** [comment]
## High Level Design
### Function: unwind
List

unwind
unwind :: String -> Record -> [Record]

This means that when given String it will return the function that will take Record and then it will give list of Record as its output.


```js
// @param {key} key by which object will be unwinded
// @param {object} object to be unwinded
unwind('hobbies', {
    name: 'alice',
    hobbies: ['Golf', 'Hacking'],
    colors: ['red', 'green'],
});
// => [
//      { name: 'alice', hobbies: 'Golf', colors: ['red', 'green'] },
//      { name: 'alice', hobbies: 'Hacking', colors: ['red', 'green'] }
// ]
```
key 
Key to determine which property of the object should be unwind
object
Object containing list under property named as key which is to unwind

unwind means separating group into multiple individual elements

Returns list of object containing the value of input key having list replaced by each element in the object.

See also unwindAll, wind.

```js
const xs = [
    { name: 'Francisco', hobbies: ['Hacking', 'Ice skating'] },
    { name: 'Adelina', hobbies: ['Juggling', 'Graffiti', 'Painting'] },
];

map(unwind('hobbies')(xs);
// => [
//     [
//         { name: 'Francisco', hobbies: 'Hacking'},
//         { name: 'Francisco', hobbies: 'Ice skating'},
//     ],
//     [
//         { name: 'Adelina', hobbies: 'Juggling'},
//         { name: 'Adelina', hobbies: 'Graffiti'},
//         { name: 'Adelina', hobbies: 'Painting'},
//     ]
// ]
// [{},{}] -> [ [{},{}], [{},{}] ]
```


### Function: unwindAll
List

unwindAll
unwindAll :: List -> Record -> [Record]

This means that when given List it will return the function that will take Record and then it will give list of Record as its output.

keys
Keys to determine which properties of the object should be unwind
object
Object containing list under properties named as keys element which is to unwind

unwind means separating group into multiple individual elements

Returns list of object containing the value of input keys having list replaced by each element in the object.

See also unwind, wind

```js
// @param {keys} keys by which object will be unwinded
// @param {object} object to be unwinded
unwindAll(['hobbies', 'colors'], {
    name: 'alice',
    hobbies: ['Golf', 'Hacking'],
    colors: ['red', 'green'],
});

// => [
//      { name: 'alice', hobbies: 'Golf', colors: 'red' },
//      { name: 'alice', hobbies: 'Golf', colors: 'green' },
//      { name: 'alice', hobbies: 'Hacking', colors: 'red' },
//      { name: 'alice', hobbies: 'Hacking', colors: 'green' },
// ]

// => [
//    [
//        { name: 'alice', hobbies: 'Golf', colors: ['red', 'green'] },  
//        { name: 'alice', hobbies: 'Hacking', colors:  ['red', 'green'] },
//    ],
//    [
//        { name: 'alice', hobbies: ['Golf', 'Hacking'], colors: 'red' },  
//        { name: 'alice', hobbies: ['Golf', 'Hacking'], colors:  'green' },
//    ]
// ]
```

### Function: wind
List

wind
wind :: String -> [Record] -> Record

This means that when given String it will return the function that will take List of Record and then it will give Record as its output.

key
Key to determine which property of the object should be wind

list
List containing element with property named as keys

wind means combining multiple individual elements into one single group

Returns object containing the value of each element with key replaced by single list.

See also unwind, unwindAll
```js
// @param {key} key by which object will be winded
// @param {object} object to be winded
wind('hobby',[
    { name: 'alice', hobby: 'Hacking' },
    { name: 'alice', hobby: 'Juggling' },
])
// => { name: 'alice', hobby: ['Hacking', 'Juggling'] }
```

**Comment by Jayesh-30** [comment]
## Detail Level Design
### Function: unwind
1. Determine key and object so as to get key which is needed to unwind over the object.
```
key <- 'hobbies'
object <- {
    name: 'alice',
    hobbies: ['Golf', 'Hacking'],
    colors: ['red', 'green']
}
```
2. Need to reduce over the list named by object's key
```
object[key] = ['Golf', 'Hacking'];
```

3. reduce method executes a reducer function on each element of the list, resulting in single output final accumulator value.

4. Initialize accumulator as empty list
```
accumulator <- [];
```

5. And by reducing over it is required to add the whole object into the accumulator
```
accumulator <- [
    object,
    object
];
```
6. In every iteration of reducer function we need to replace the object[key] ie ['Golf', 'Hacking'] with each element of the list ie 'Golf' and 'Hacking'.
```
In first iteration
accumulator <- [
    {
        name: 'alice',
        hobbies: 'Golf',
        colors: ['red', 'green']
    },
];
In second iteration
accumulator <- [
    {
        name: 'alice',
        hobbies: 'Golf',
        colors: ['red', 'green']
    },
    {
        name: 'alice',
        hobbies: 'Hacking',
        colors: ['red', 'green']
    }
];
```
Similarly, at end of it all the object[key] will be unwinded into its own individual object.

7. Return the accumulator as final output.
```
return accumulator
```

### Function: unwindAll
1. Determine keys and object so as to get keys which is needed to unwind over the object.
```
keys <- ['hobbies','colors']
object <- {
    name: 'alice',
    hobbies: ['Golf', 'Hacking'],
    colors: ['red', 'green']
}
```
2. Need to map over the keys so as to obtain object[key]
```
hobbies <- ['Golf', 'Hacking']
colors <- ['red', 'green']
```
3. While mapping over keys we need to unwind the object with key of each iteration.
```
result_1 <- unwind('hobbies', object)
result_2 <- unwind('colors', object)
...
result_n
```
4. Initialize the result with empty list
```
result <- [];
```
5. Add all the result_1, result_2, ... result_n into result.
```
result <- [result_1, result_2, ... result_n];
```
6. In every iteration of mapping over keys.
```
In first iteration
result <- [
    unwind('hobbies', object)
];
In second iteration
result <- [
    unwind('hobbies', object),
    unwind('colors', object)
];
```
Similarly, at end of it all the keys will be unwinded into its own individual object and added to the result list.

7. Return the result as final output.
```
return result
```

### Function: wind
1. Determine key and list so as to get key which is needed to wind the list into single object.
```
key <- 'hobby'
list <- [
    { name: 'alice', hobby: 'Hacking' },
    { name: 'alice', hobby: 'Juggling' },
]
```
2. Initialize an empty object with value of elemnent of the list
```
object <- {
    name: 'alice'
}
```
3. Initialize an empty list of list[key]
```
hobby <- []
```
4. Iterate over the list and then add each element to the empty list.

5. While iterating over list we need to wind the object with key ie hobby.
```
In first iteration
hobby <- ['Hacking']
In second iteration
hobby <- ['Hacking', 'Juggling']
```
Similarly, at end of it all the list element will be winded into single object.

6. Add this list to the object under object[key]
```
object <- {
    name: 'alice', hobby: ['Hacking', 'Juggling']
}
```
7. Return the object as final output.
```
return object
```

**Comment by CrossEye** [comment]
This looks great..  I can't wait to see the actual code!

**Comment by customcommander** [comment]
Closing as implemented in #3125
